public/modules/custom/helfi_ahjo/src/Services/AhjoService.php

```diff
@@ -0,0 +1,291 @@
+   *   Data for batch.
+   *
+   * @return void
+   */
+  public function createTaxonomyBatch($data) {
+    if (!is_array($data)) {
+      $data = Json::decode($data);
+    }
+
+    $operations = [];
+
+    $operations[] = $this->setAllBatchOperations($data);
+
+    $batch = [
+      'operations' => $operations,
+      'finished' => $this->createTaxonomyTermsBatchFinished(),
+      'title' => 'Performing an operation',
+      'init_message' => 'Please wait',
+      'progress_message' => 'Completed @current from @total',
+      'error_message' => 'An error occurred',
+    ];
+
+    batch_set($batch);
+
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function addToCron($data, $queue, $parentId = NULL) {
+    if (!is_array($data)) {
+      $data = Json::decode($data);
+    }
+
+    foreach ($data as $section) {
+      $section['parentId'] = $parentId ?? 0;
+      $queue->createItem($section);
+
+      if (isset($section['OrganizationLevelBelow'])) {
+        $this->addToCron($section['OrganizationLevelBelow'], $queue, $section['ID']);
+      }
+    }
+    $this->syncTaxonomyTermsChilds();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function syncTaxonomyTermsChilds() {
+    $terms = $this->entityTypeManager
+      ->getStorage('taxonomy_term')
+      ->loadByProperties(['vid' => 'sote_section']);
+    foreach ($terms as $item) {
+      if (!isset($item->field_external_parent_id->value)
+        || $item->field_external_parent_id->value == NULL
+        || $item->field_external_parent_id->value == '0') {
+        continue;
+      }
+      $loadByExternalId = $this->entityTypeManager->getStorage('taxonomy_term')->loadByProperties([
+        'vid' => 'sote_section',
+        'field_external_id' => $item->field_external_parent_id->value,
+      ]);
+
+      $item->set('parent', reset($loadByExternalId)->tid->value);
+      $item->save();
+
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function showDataAsTree($excludedByTypeId = [], $organization = 0, $maxDepth = 0) {
+    $terms = $this->entityTypeManager->getStorage('taxonomy_term')->loadTree('sote_section', $organization, $maxDepth);
+
+    $tree = [];
+    foreach ($terms as $tree_object) {
+      $term = $this->entityTypeManager->getStorage('taxonomy_term')->load($tree_object->tid);
+      if (in_array($term->field_type_id, $excludedByTypeId)) {
+        continue;
+      }
+      $this->taxonomyUtils->buildTree($tree, $tree_object, 'sote_section');
+    }
+
+    return $tree;
+  }
+
+  public function createTaxonomyTermsBatch($data, &$context) {
+    $message = 'Creating taxonomy terms...';
+
+    $loadByExternalId = $this->entityTypeManager->getStorage('taxonomy_term')->loadByProperties([
+      'vid' => 'sote_section',
+      'field_external_id' => $data['ID'],
+    ]);
+
+    if (count($loadByExternalId) == 0) {
+      $term = Term::create([
+        'name' => $data['Name'],
+        'vid' => 'sote_section',
+        'field_external_id' => $data['ID'],
```

Comment:
check in $context['importedTids'][$data['parentId']] and set the parent_id to that tid